---
title: "ChromENVEE: Chromatin ENVironment and Expression at Enhancer"
author: "Coulée Manon"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: true

vignette: >
  %\VignetteIndexEntry{ChromENVEE: Chromatin ENVironment and Expression at Enhancer}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Abstract

# Citation

# Introduction

ChromENVEE is a package developed to study chromatin state without Hi-C data.

This package implements functions to associate genes with enhancers, to define the chromatin environment of gene from genomic data (e.g., ChromHMM output or a bed file). Several visualization functions are available to summarize the distribution of chromatin states, characterize genes associated to enhancers and also estimate the chromatin environment of genes.

This package is available for R version >= 3.6.

```{r setup}
# Loading package
library(ChromENVEE)
```

# Initialization of data

`colorTable` is a dataframe which contains informations about chromatin state number (`stateNumber`), chromatin state name (`stateName`) and chromatin state color (`colorValue`). This table is necessary for plot generation. `colorValue` accepted as value hex code and/or color name code.

```{r}
data(colorTable)
```

```{r echo = FALSE, fig.height = 10,fig.asp = 0.6}
library(grid)
library(gridExtra)
tt = ttheme_minimal(
  core = list(bg_params = list(fill = c(rep("white",36),colorTable$colorValue), col = 1))
)
grid.table(colorTable, theme = tt)
```

`genomeFile` is a dataframe which contains informations about mouse reference genome.

It is generated from an annotation bed file, in the case of this study we used the Ensembl annotation. `genomeFile` should contain information such as chromosome (chr), gene position (start and end), strand information (strand) and gene name (gene_ENS). Score information is suggested but not mandatory.

```{r}
data(genomeFile)
```
```{r echo = FALSE}
head(genomeFile)
```

`chromatinState` is a dataframe which contains informations about chromatin state.

It is generated with the output of the ChromHMM tool. `chromatinState` should contain information such as chromosome (chr), genomic regions (start and end), chromatin state information (state and state_name) and sample name (name).

```{r}
data(chromatinState)
```
```{r echo = FALSE}
head(chromatinState)
```

# Distribution of chromatin state in the genome

We want to know the distribution of chromatin state in the genome.

`plotChromatinState` calculates the percentage of each chromatin state according to the length of the genome used. We obtain a dataframe with the percentage of coverage for each chromatin state. It's possible to plot the result in `.png` file with the argument `plot = TRUE`. If you have a list of dataframe, it's possible to merge all the dataframe in an unique merge dataframe and in an unique plot with the argument `merge = TRUE`.

```{r}
summary_chromatin_state = plotChromatinState(chromatinState, merge = TRUE, plot = FALSE,
colorTable = colorTable, filename = "")
```
```{r}
head(summary_chromatin_state)
```
![](figures/chromatin_state_distribution.png){width=100%}

# Annotation of enhancer

We want to associate at each enhancer, all genes regulated by the enhancer. We focus on enhancer chromatin state (in this study, we have 4 types of enhancer : bivalent enhancer (EnhBiv), genic enhancer (EnhG), active enhancer (EnhA) and weak enhancer (EnhWk)).

`listTableEnhancer` is a GRanges object or a list of GRanges object (producted by [`GenomicRanges`](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) package). Like chromatinState dataframe, `listTableEnhancer` should contain gene and chromatin state informations. Sample name (sample_name) is mandatory if you want to compare enhancer annotation (see [Enhancer annotation comparison](#enhancerlist)).

```{r}
data(listTableEnhancer)
```
```{r echo = FALSE}
listTableEnhancer[[1]]
```

## Annotated enhancer binding to enhancer position

To estimate which gene is regulated by enhancer, we estimated as enhancer-associated genes, all genes in an interval around enhancer.
`enhancerAnnotation()` uses a `GRanges` object.

The function takes few minutes to process depending on the length of your enhancer table. It's possible to multithread the function with the `nCore` parameter. For each enhancer position, we get two informations, distance between gene and enhancer (in bp) and gene associate.


```{r}
table_enhancer_gene = enhancerAnnotation(listTableEnhancer[[1]], genome = genomeFile,
interval = 500000, nCore = 1)
```

### Number of gene associate at the enhancer

We want to know the distribution of genes associated at each enhancer using `plotGeneAssociation`.
```{r, fig.width = 10,fig.asp = 0.6}
plotGeneAssociation(table_enhancer_gene, all = FALSE)
```

## Associated gene expression to enhancer

`geneExpression` is a dataframe which contains gene expression level information.

It is generated with RNAseq gene expression analysis. `geneExpression` should contain informations like chromosome (chr), gene position (start and end), gene name (gene_ENS), strand information (strand), level of gene expression (gene_expression). Score is not necessary for the analysis. For gene name, you must use the same gene name you used to generate `genomeFile` dataframe because we use the annotation to associate the gene-enhancer pair with the expression.

```{r}
data(geneExpression)
```
```{r echo = FALSE}
head(geneExpression)
```

We want to associate the level of gene expression at each gene-enhancer pair are estimated with `enhancerAnnotation` function.

According to `geneExpression` dataframe, it's possible that gene-enhancer pair has not expression level, in this case, function return NA value.

```{r}
table_enhancer_gene_expression = enhancerExpression(table_enhancer_gene,
geneExpressionTable = geneExpression)
```
```{r echo = FALSE}
head(table_enhancer_gene_expression)
```

## Visualization of enhancer annotation
### Expression of gene associated to enhancer according to their distance

We generated plot to estimate the level of gene expression according to the distance between gene and enhancer using `plotDistanceExpression`. The distance is estimated with `limit` argument and clusterized in 6 distance groups like the following plot.

```{r, fig.width = 10,fig.asp = 0.6}
plotDistanceExpression(table_enhancer_gene_expression, colorTable = colorTable,
limit = 500000)
```

### Distribution of gene according to distance between gene and enhancer

We generated plot to estimate the distribution of gene according to the distance between gene and enhancer using `plotGeneDistance`. The distance is estimated with `limit` argument and clusterized in 6 distance groups like the following plot.
```{r, fig.width = 10,fig.asp = 0.3}
plotGeneDistance(table_enhancer_gene_expression, limit = 500000, xlab = "",
ylab = "distance enhancer-gene (bp)")
```

### Expression of gene associated to enhancer

We generated plot with the distribution of gene expression associated at enhancer region using `plotEnhancerExpression`. It's possible to rescale the plot with `scale` argument ('none','log10' and 'log2' are accepted).
```{r, fig.width = 10,fig.asp = 0.6}
plotEnhancerExpression(table_enhancer_gene_expression, scale = "log10",
colorTable = colorTable, ylab = "gene expression log10(CPM)")
```

# Enhancer annotation comparison <a name="enhancerlist"></a>

It's possible to compare different categories of enhancers. To do this, it's necessary to use a list of GRanges objects, each containing data like those in `listTableEnhancer`. Unlike the individual analysis, each GRanges object in the list requires sample information (sample_name).

The first step is associate gene with each enhancer using `enhancerAnnotation()` on the list of enhancer. After gene association, we associated the gene expression using `enhancerExpression()`. In the case of this study, all enhancer categories are from the same cell type, we also used the same `geneExpression` dataframe.


```{r}
list_table_enhancer_gene = lapply(listTableEnhancer, enhancerAnnotation,
genome = genomeFile,interval = 500000, nCore = 1)
listTableEnhancerGeneExpression = lapply(list_table_enhancer_gene, enhancerExpression,
geneExpressionTable = geneExpression)
```

This process take many time. To reduce time, you can load the `listTableEnhancerGeneExpression` data to process the following analyses.

`data(listTableEnhancerGeneExpression)`

### Number of gene associate at the enhancer

We want to know the distribution of genes associated at each enhancer using `plotGeneAssociation`. `all = TRUE` parameter is used to compile all enhancer tables in same '.png' file.

```{r, fig.width = 10,fig.asp = 0.6}
plotGeneAssociation(listTableEnhancerGeneExpression, all = TRUE)
```

### Expression of gene associated to enhancer according to their distance

We generated plot to estimate the level of gene expression according to the distance between gene and enhancer using `plotDistanceExpression`. The distance is estimated with `limit` argument and clusterized in 6 distance groups like the following plot. In the case of list analysis, the function showed the average of the expression associate to each enhancer.

```{r, fig.width = 10,fig.asp = 0.6}
plotDistanceExpression(listTableEnhancerGeneExpression, colorTable = colorTable,
limit = 500000)
```

### Distribution of gene according to distance between gene and enhancer

We generate plot to estimated the distribution of gene according to the distance between gene and enhancer using `plotGeneDistance`. The distance is estimated with `limit` argument and clusterized in 6 distance groups like the following plot.
```{r, fig.width = 10,fig.asp = 0.6}
plotGeneDistance(listTableEnhancerGeneExpression, limit = 500000,
xlab = "", ylab = "distance enhancer-gene (bp)")
```

### Expression of gene associated to enhancer

We generate plot with the distribution of gene expression associated at enhancer region using `plotEnhancerExpression`. It's possible to rescale the plot with `scale` argument ('none', 'log10' and 'log2' are accepted).
```{r, fig.width = 10,fig.asp = 0.6}
plotEnhancerExpression(listTableEnhancerGeneExpression, scale = "log10",
colorTable = colorTable, ylab = "gene expression log10(CPM)")
```

# Chromatin state gene environment

We want to study the chromatin environment of gene. To do this, we need gene information data (`geneExpression`) and chromatin state data (`chromatinState`).

```{r}
data(geneExpression)
data(chromatinState)
```

## Coverage of chromatin state in environment binding to TSS regions

`geneEnvironment()` is a function able to estimate the chromatin state environment of gene. For this, we estimated the size of the environment around gene TSS with `interval` parameter. For each gene, we obtain information about the coverage of each chromatin state (`state` parameter) in the environment.

`geneEnvironment()` may take few minutes depending on the number of genes analyzed.

```{r}
table_overlapping = geneEnvironment(geneExpression, chromatinState,
stateOrder = unique(colorTable$stateName), interval = 3000)
rownames(table_overlapping) = table_overlapping$gene_ENS
```
```{r echo = FALSE}
head(table_overlapping)
```

## Predominant state in environment binding to TSS regions

`predominantState()` estimates the predominant chromatin state in the gene environment.
The function estimates as predominant the chromatin state with the highest coverage in the environment. Genes are clusterized according to their chromatin state using [`umap`](https://cran.r-project.org/web/packages/umap/index.html) package. The function returns a dataframe with information about the predominant chromatin state and UMAP dimension.

```{r}
result_umap = predominantState(table_overlapping, state = unique(colorTable$stateName),
header = unique(colorTable$stateName), neighbors = 32, metric = "euclidean", dist = 0.5)
```
```{r echo = FALSE}
head(result_umap)
```

It's an exemple of UMAP representation to visualize the predominant chromatin state in each gene. Each point corresponds to a gene and each is colored according to its predominant chromatin state

![](figures/umap_chromatin_state.png){width=100%}

# Session Information

Here is the output of `sessionInfo()` on the system on which this document was compiled:

```{r echo = FALSE}
sessionInfo()
```

# References

Ernst J, Kellis M. ChromHMM: automating chromatin-state discovery and characterization. Nature Methods, 9:215-216, 2012

Papier scientifique associé

McInnes, Leland, and John Healy. "UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction." arXiv:1802.03426.

Lawrence M, Huber W, Pagès H, Aboyoun P, Carlson M, Gentleman R, Morgan M, Carey V (2013). “Software for Computing and Annotating Genomic Ranges.” PLoS Computational Biology, 9. doi: 10.1371/journal.pcbi.1003118, http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118.
