---
title: "ChromENVEE: Chromatin ENVironment and Enhancer Expression"
author: "CoulÃ©e Manon"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: true


vignette: >
  %\VignetteIndexEntry{ChromENVEE: Chromatin ENVironment and Enhancer Expression}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Abstract

# Citation

# Introduction

ChromENVEE is a package developped to studying chromatin state without Hi-C data.

This package implements functions to associated genes with enhancers, define the chromatin environment of the gene from genomic data (e.g., ChromHMM output or a bed file). Several visualization functions are available to summarize the distribution of chromatin states, characterize genes associated with enhancers and also estimate the chromatin environment of genes.

```{r setup}
# Loading package
library(ChromENVEE)
```

# Initialization of data

Created initial vector contains informations about chromatin state number `stateNumber`, chromatin state name `stateName` and chromatin state color `colorValue` (used for plot generation).

```{r}
stateNumber = c("U1","U2","U3","U4","U5","U6","U7","U8","U9","U10","U11","U12","U13","U14","U15",
"U16","U17","U18")
stateName = c("TSSA","TSSFlnk","TSSFlnkD","Tx","TxWk","EnhG","EnhG","EnhA","EnhWk","ZNFRpts","Het",
"TssBiv","EnhBiv","ReprPC","ReprPCWk","Quies","Quies","Quies")
colorValue = c("#B71C1C","#E65100","#E65100","#43A047","#1B5E20","#99FF66","#99FF66","#F5B041",
"#FFEB3B","#48C9B0","#B39DDB","#880E4F","#666633","#424949","#7B7D7D","#D0D3D4","#D0D3D4","#D0D3D4")
```

`genomeFile` is a data frame contains informations about mouse reference genome.

It is generated from bed file, in the case of this study, we used Ensembl annotation. `genomeFile`
required to contains informations like chromosome (chr), gene position (start and end), strand
information (strand) and gene name (gene_ENS). Score informations is suggested but not required.

```{r}
data(genomeFile)
```
```{r echo = FALSE}
head(genomeFile)
```

`chromatinState` is a data frame contains informations about chromatin state.

It is generated with the output of ChromHMM tools. `chromatinState` required to contains
informations like chromosome (chr), genomic regions (start and end), chromatin state information
(state and state_name) and sample name (name)


```{r}
data(chromatinState)
```
```{r echo = FALSE}
head(chromatinState)
```
# Distribution of chromatin state in the genome

We are interested to know the distribution of chromatin state in the genome.

`plotChromatinState` calculates the percentage of each chromatin state in function the length of
the genome used. We obtains a data frame with the percentage of coverage for each chromatin state.
It's possible to plot the result in `.png` file with the argument `plot = TRUE`. If you have a list
of data frame, it's possible to merge all the data frame in merge data frame and in unique plot
with `merge = TRUE` argument.

```{r}
summary_chromatin_state = plotChromatinState(chromatinState, stateName = stateName,
stateNumber = stateNumber, merge = TRUE, plot = FALSE, color = colorValue, filename = "")
```
```{r}
head(summary_chromatin_state)
```
![](figures/chromatin_state_distribution.png){width=100%}


# Annotation of enhancer

We are interested to associated at each enhancer, genes regulated by the enhancer. We focused on
enhancer chromatin state (in this study, we have 4 tpe of enhancer : bivalent enhancer (EnhBiv),
genic enhancer (EnhG), active enhancer (EnhA) and weak enhancer (EnhWk)).

`listTableEnhancer` is a GRanges object or a list of GRanges object (producted by [`GenomicRanges`](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) package). Like chromatinState data frame, `listTableEnhancer` required gene information and chromatin state information. Sample name (sample_name) is required if you want compared enhancer annotation (see Enhnacer annotation comparison part).

```{r}
data(listTableEnhancer)
```
```{r echo = FALSE}
listTableEnhancer[[1]]
```
## Annotated enhancer binding to enhancer position

To estimated which gene is regulated by enhancer, we estimated that genes associated enhancer, all
TSS genes in interval around enhancer.
`enhancerAnnotation()` uses a `GRanges` object.

The function take few minutes to process in function the length of your enhancer table. It's possible to multithread the job with the `nCore` parameter. For each enhancer position, we obtains two informations, the distance between gene and enhancer (in bp) and the gene.


```{r}
table_enhancer_gene = enhancerAnnotation(listTableEnhancer[[1]],genome = genomeFile,
interval = 500000, nCore = 1)
```

### Number of gene associate at the enhancer
We want to know the distribution of genes associated at each enhancer.
```{r, fig.width = 7,fig.asp = 0.6}
plotGeneAssociation(table_enhancer_gene, all = FALSE)
```

## Associated gene expression to enhancer

`geneExpression` is a data frame contains gene expression level information.

It is generated with RNAseq gene expression analysis. `geneExpression` required to contains
informations like chromosome (chr), gene position (start and end), gene name (gene_ENS), strand
information (strand), level of gene expression (gene_expression). Score is not required for the
analysis. For gene name, you need to used the same gene name that you used to generated `genomeFile`
data frame because we used the annotation to associated the couple gene-enhancer with the expression.

```{r}
data(geneExpression)
```
```{r echo = FALSE}
head(geneExpression)
```

We associated the level of gene expression at each gene-enhancer couple estimate with `enhancerAnnotation` function.

According to `geneExpression` data frame, it's possible that gene-enhancer couple has not expression level, in this case, we obtains NA value.

```{r}
table_enhancer_gene_expression = enhancerExpression(table_enhancer_gene,
  geneExpressionTable = geneExpression)
```
```{r echo = FALSE}
head(table_enhancer_gene_expression)
```
