---
title: "ChromENVEE: Chromatin ENVironment and Enhancer Expression"
author: "Coulée Manon"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: true

vignette: >
  %\VignetteIndexEntry{ChromENVEE: Chromatin ENVironment and Enhancer Expression}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Abstract

# Citation

# Introduction

ChromENVEE is a package developped to studying chromatin state without Hi-C data.

This package implements functions to associated genes with enhancers, define the chromatin environment of the gene from genomic data (e.g., ChromHMM output or a bed file). Several visualization functions are available to summarize the distribution of chromatin states, characterize genes associated with enhancers and also estimate the chromatin environment of genes.

This package is available for R version >= 3.6.

```{r setup}
# Loading package
library(ChromENVEE)
```

# Initialization of data

`colorTable` is a data frame contains informations about chromatin state number `stateNumber`, chromatin state name `stateName` and chromatin state color `colorValue` (used for plot generation).

```{r}
data(colorTable)
```

```{r echo = FALSE}
library(grid)
library(gridExtra)
tt = ttheme_minimal(
  core = list(bg_params = list(fill = c(rep("white",36),colorTable$colorValue), col = 1))
)
grid.table(colorTable, theme = tt)
```

`genomeFile` is a data frame contains informations about mouse reference genome.

It is generated from bed file, in the case of this study, we used Ensembl annotation. `genomeFile`
required to contains informations like chromosome (chr), gene position (start and end), strand
information (strand) and gene name (gene_ENS). Score informations is suggested but not required.

```{r}
data(genomeFile)
```
```{r echo = FALSE}
head(genomeFile)
```

`chromatinState` is a data frame contains informations about chromatin state.

It is generated with the output of ChromHMM tools. `chromatinState` required to contains
informations like chromosome (chr), genomic regions (start and end), chromatin state information
(state and state_name) and sample name (name).


```{r}
data(chromatinState)
```
```{r echo = FALSE}
head(chromatinState)
```

# Distribution of chromatin state in the genome

We are interested to know the distribution of chromatin state in the genome.

`plotChromatinState` calculates the percentage of each chromatin state in function the length of
the genome used. We obtains a data frame with the percentage of coverage for each chromatin state.
It's possible to plot the result in `.png` file with the argument `plot = TRUE`. If you have a list
of data frame, it's possible to merge all the data frame in merge data frame and in unique plot
with `merge = TRUE` argument.

```{r}
summary_chromatin_state = plotChromatinState(chromatinState, merge = TRUE, plot = FALSE,
colorTable = colorTable, filename = "")
```
```{r}
head(summary_chromatin_state)
```
![](figures/chromatin_state_distribution.png){width=100%}

# Annotation of enhancer

We are interested to associated at each enhancer, genes regulated by the enhancer. We focused on
enhancer chromatin state (in this study, we have 4 type of enhancer : bivalent enhancer (EnhBiv),
genic enhancer (EnhG), active enhancer (EnhA) and weak enhancer (EnhWk)).

`listTableEnhancer` is a GRanges object or a list of GRanges object (producted by [`GenomicRanges`](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) package). Like chromatinState data frame, `listTableEnhancer` required gene information and chromatin state information. Sample name (sample_name) is required if you want compared enhancer annotation (see [Enhancer annotation comparison](#enhancerlist)).

```{r}
data(listTableEnhancer)
```
```{r echo = FALSE}
listTableEnhancer[[1]]
```

## Annotated enhancer binding to enhancer position

To estimated which gene is regulated by enhancer, we estimated that genes associated enhancer, all
TSS genes in interval around enhancer.
`enhancerAnnotation()` uses a `GRanges` object.

The function take few minutes to process in function the length of your enhancer table. It's possible to multithread the job with the `nCore` parameter. For each enhancer position, we obtains two informations, the distance between gene and enhancer (in bp) and the gene.


```{r}
table_enhancer_gene = enhancerAnnotation(listTableEnhancer[[1]],genome = genomeFile,
interval = 500000, nCore = 1)
```

### Number of gene associate at the enhancer

We want to know the distribution of genes associated at each enhancer using `plotGeneAssociation`.
```{r, fig.width = 10,fig.asp = 0.6}
plotGeneAssociation(table_enhancer_gene, all = FALSE)
```

## Associated gene expression to enhancer

`geneExpression` is a data frame contains gene expression level information.

It is generated with RNAseq gene expression analysis. `geneExpression` required to contains
informations like chromosome (chr), gene position (start and end), gene name (gene_ENS), strand
information (strand), level of gene expression (gene_expression). Score is not required for the
analysis. For gene name, you need to used the same gene name that you used to generated `genomeFile`
data frame because we used the annotation to associated the couple gene-enhancer with the expression.

```{r}
data(geneExpression)
```
```{r echo = FALSE}
head(geneExpression)
```

We associated the level of gene expression at each gene-enhancer couple estimate with `enhancerAnnotation` function.

According to `geneExpression` data frame, it's possible that gene-enhancer couple has not expression level, in this case, we obtains NA value.

```{r}
table_enhancer_gene_expression = enhancerExpression(table_enhancer_gene,
geneExpressionTable = geneExpression)
```
```{r echo = FALSE}
head(table_enhancer_gene_expression)
```

## Profile of enhancer annotation
### Distance gene-enhancer according to their expression

We generated plot to estimated the level of gene expression according to the distance between gene and enhancer using `plotDistanceExpression`. The distance is estimated with `limit` argument and clusterized in 6 distance groups like the following plot.

```{r, fig.width = 10,fig.asp = 0.6}
plotDistanceExpression(table_enhancer_gene_expression, colorTable = colorTable,
limit = 500000)
```

### Distance gene-enhancer

We generated plot to estimated the distribution of gene according to the distance between gene and enhancer using `plotGeneDistance`. The distance is estimated with `limit` argument and clusterized in 6 distance groups like the following plot.
```{r, fig.width = 10,fig.asp = 0.3}
plotGeneDistance(table_enhancer_gene_expression, limit = 500000, xlab = "",
ylab = "distance enhancer-gene (bp)")
```

### Enhancer expression

We generated plot with the distribution of gene expression associated at enhancer region using `plotEnhancerExpression`. It's possible to rescale plot with `scale` argument ('none','log10' and 'log2' are accepted).
```{r, fig.width = 10,fig.asp = 0.6}
plotEnhancerExpression(table_enhancer_gene_expression, scale = "log10",
colorTable = colorTable, ylab = "gene expression log10(CPM)")
```

# Enhancer annotation comparison <a name="enhancerlist"></a>

It's possible to compared different categories of enhancer. For that it's necessary to used a list of GRanges object each contains data like `listTableEnhancer` data. Contrary to individual analysis, each GRanges object in the list required sample information (sample_name).

The first step is associated gene to each enhancer using `enhancerAnnotation()` on the list of enhancer. After the gene association, we associated the gene expression using `enhancerExpression()`. In the case of this study, all enhancer categories come from same cell type, we also used the same `geneExpression` data frame.


```{r}
list_table_enhancer_gene = lapply(listTableEnhancer, enhancerAnnotation,
genome = genomeFile,interval = 500000, nCore = 1)
listTableEnhancerGeneExpression = lapply(list_table_enhancer_gene, enhancerExpression,
geneExpressionTable = geneExpression)
```

This process take many time.

To reduce time, you can load `listTableEnhancerGeneExpression` data to process the next analysis.


`data(listTableEnhancerGeneExpression)`

### Number of gene associate at the enhancer

We want to know the distribution of genes associated at each enhancer using `plotGeneAssociation`. `all = TRUE` parameter is used to compiled all enhancer table in same file.

```{r, fig.width = 10,fig.asp = 0.6}
plotGeneAssociation(listTableEnhancerGeneExpression, all = TRUE)
```


### Distance gene-enhancer according to their expression

We generate plot to estimated the level of gene expression according to the distance between gene and enhancer using `plotDistanceExpression`. The distance is estimated with `limit` argument and clusterized in 6 distance groups like the following plot. In the case of list analysis, the function showed the mean of expression.
```{r, fig.width = 10,fig.asp = 0.6}
plotDistanceExpression(listTableEnhancerGeneExpression, colorTable = colorTable,
limit = 500000)
```


### Distance gene-enhancer

We generate plot to estimated the distribution of gene according to the distance between gene and enhancer using `plotGeneDistance`. The distance is estimated with `limit` argument and clusterized in 6 distance groups like the following plot.
```{r, fig.width = 10,fig.asp = 0.6}
plotGeneDistance(listTableEnhancerGeneExpression, limit = 500000,
xlab = "", ylab = "distance enhancer-gene (bp)")
```

### Enhancer expression

We generate plot with the distribution of gene expression associated at enhancer region using `plotEnhancerExpression`. It's possible to rescale plot with `scale` argument ('none','log10' and 'log2' are accepted).
```{r, fig.width = 10,fig.asp = 0.6}
plotEnhancerExpression(listTableEnhancerGeneExpression, scale = "log10",
colorTable = colorTable, ylab = "gene expression log10(CPM)")
```

# Gene environment

We are interested to studying the chromatin environment of gene. For that we need gene information data (`geneExpression`) and chromatin state data (`chromatinState`).

```{r}
data(geneExpression)
data(chromatinState)
```

## Coverage of chromatin state in environment binding to TSS regions

`geneEnvironment` is a function able to estimate the chromatin state environment of gene. For that, we estimated the environment size around gene TSS with `interval` parameter. For each gene, we obtains informations about the coverage of each chromatin state (`state` parameter) in the environment.

`geneEnvironment` may take few minutes in function the number of genes analyzed.

```{r}
table_overlapping = geneEnvironment(geneExpression, chromatinState,
stateOrder = unique(colorTable$stateName), interval = 3000)
rownames(table_overlapping) = table_overlapping$gene_ENS
```
```{r echo = FALSE}
head(table_overlapping)
```

## Predominant state in environment binding to TSS regions

`predominentState` able to estimated the predominent chromatin state in the environment of the gene.
The function estimate as predominent the chromatin state with the higher coverage in the environment. Genes are clusterized in function their chromatin state using `umap` package. The function return a data frame with information about the predominent chromatin state and UMAP dimension.
```{r}
result_umap = predominentState(table_overlapping, state = unique(colorTable$stateName),
header = unique(colorTable$stateName), neighbors = 32, metric = "euclidean", dist = 0.5)
```
```{r echo = FALSE}
head(result_umap)
```

It's an exemple of UMAP representation to visualized the predominent chromatin state in each gene.
![](figures/umap_chromatin_state.png){width=100%}

# Session Information

Here is the output of `sessionInfo()` on the system on which this document was compiled:

```{r echo = FALSE}
sessionInfo()
```

# References

Ernst J, Kellis M. ChromHMM: automating chromatin-state discovery and characterization. Nature Methods, 9:215-216, 2012

Papier scientifique associé

McInnes, Leland, and John Healy. "UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction." arXiv:1802.03426.

Lawrence M, Huber W, Pagès H, Aboyoun P, Carlson M, Gentleman R, Morgan M, Carey V (2013). “Software for Computing and Annotating Genomic Ranges.” PLoS Computational Biology, 9. doi: 10.1371/journal.pcbi.1003118, http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118.
